app/ui/browser-modern/sagas/external-sagas.js

```diff
@@ -40,10 +41,16 @@ function* maximizeWindow() {
 }
 
 function* closeWindow() {
+  // Wait for any additional operations to finish before closing the window.
+  // Add your own here if needed.
+  yield take(EffectTypes.NOTIFY_BROWSER_WINDOW_APP_STATE_DESTROYED);
```

Comment:
Yes, this is a situation where some stuff needs to happen before a browser window closes. The logic for what happens is defined somewhere else, since I wouldn't like polluting sagas with unrelated application logic. The `take` operations make sure other things happen before continuation is allowed.

To answer your questions: 
1. A window can indeed deny its closing now with these changes, since requests for closing windows need to go through the renderer process and through this saga first. As the comments imply, the logic for the expected state / actions that need to happen beforehand can be added here.

2. A window can decide to do whatever it wants before its closing. In fact, this is already the case: when a window is closed manually, it is removed from the session storage. As another example, in the next saga in this file, when a window is reloaded, it requests that its app state is restored from the session.